sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala

```diff
@@ -1658,6 +1659,43 @@ class Dataset[T] private[sql](
   def groupByKey[K](func: MapFunction[T, K], encoder: Encoder[K]): KeyValueGroupedDataset[K, T] =
     groupByKey(func.call(_))(encoder)
 
+
+  /**
+   * Aggregates the elements of this Dataset in a multi-level tree pattern.
+   *
+   * @param depth suggested depth of the tree (default: 2)
+   */
+  private[spark] def treeAggregate[U : Encoder : ClassTag](zeroValue: U)(
+      seqOp: (U, T) => U,
+      combOp: (U, U) => U,
+      depth: Int = 2): U = {
+    require(depth >= 1, s"Depth must be greater than or equal to 1 but got $depth.")
+    val sparkContext = sparkSession.sparkContext
+    val copiedZeroValue = Utils.clone(zeroValue, sparkContext.env.closureSerializer.newInstance())
+    if (rdd.partitions.length == 0) {
+      copiedZeroValue
+    } else {
+      val aggregatePartition =
+        (it: Iterator[T]) => it.aggregate(zeroValue)(seqOp, combOp)
+      var partiallyAggregated: Dataset[U] = mapPartitions(it => Iterator(aggregatePartition(it)))
```

Comment:
Why can't we call `rdd.treeAggregate` directly?

Reply:
I think we can. Because `treeAggregate` is applied on the domain object, I'm wondering if direct call on `rdd.treeAggregate` will be more efficient or not.

Although there is serialization happened between `mapPartitions` and `repartition`, `rdd.treeAggregate` also needs to serialize domain object during shuffling. Currently this is based on that serialization by SQL catalyst is more efficient than rdd.

Reply:
I doubt how much perf improvement we can get here, at least using `RDD.treeAggregate` can make it much simpler. Maybe we should benchmark.

Reply:
Since this benchmark should be performed on a cluster not a local machine, I will try to run a benchmark later once I set up one.